srml/evm/src/lib.rs

```diff
@@ -0,0 +1,247 @@
+	fn execute(
+		_address: H160,
+		_input: &[u8],
+		_target_gas: Option<usize>
+	) -> Option<core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>> {
+		None
+	}
+}
+
+/// EVM module trait
+pub trait Trait: system::Trait + timestamp::Trait {
+	/// Calculator for current gas price.
+	type FeeCalculator: FeeCalculator;
+	/// Convert account ID to H160;
+	type ConvertAccountId: ConvertAccountId<Self::AccountId>;
+	/// Currency type for deposit and withdraw.
+	type Currency: Currency<Self::AccountId>;
+	/// The overarching event type.
+	type Event: From<Event> + Into<<Self as system::Trait>::Event>;
+	/// Precompiles associated with this EVM engine.
+	type Precompiles: Precompiles;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		Accounts get(fn accounts) config(): map H160 => Account;
+		AccountStorages: double_map H160, blake2_256(H256) => H256;
+	}
+}
+
+decl_event!(
+	/// EVM events
+	pub enum Event {
+		/// Ethereum events from contracts.
+		Log(Log),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn deposit_balance(origin, value: BalanceOf<T>) -> Result {
+			let sender = ensure_signed(origin)?;
+
+			T::Currency::withdraw(
+				&sender,
+				value,
+				WithdrawReason::Reserve,
+				ExistenceRequirement::KeepAlive,
+			)?;
+
+			let bvalue = U256::from(UniqueSaturatedInto::<u64>::unique_saturated_into(value));
+			let address = T::ConvertAccountId::convert_account_id(sender);
+			Accounts::mutate(&address, |account| {
+				account.balance += bvalue;
+			});
+
+			Ok(())
+		}
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn withdraw_balance(origin, value: BalanceOf<T>) -> Result {
+			let sender = ensure_signed(origin)?;
+			let address = T::ConvertAccountId::convert_account_id(sender.clone());
+			let bvalue = U256::from(UniqueSaturatedInto::<u64>::unique_saturated_into(value));
+
+			if Accounts::get(&address).balance < bvalue {
+				return Err("Not enough balance to withdraw")
+			}
+
+			Accounts::mutate(&address, |account| {
+				account.balance -= bvalue;
+			});
+
+			T::Currency::deposit_creating(&sender, value);
+
+			Ok(())
+		}
+
+		#[weight = SimpleDispatchInfo::FixedNormal(10_000)]
+		fn call(origin, target: H160, input: Vec<u8>, value: U256, gas_limit: u32) -> Result {
+			let sender = ensure_signed(origin)?;
+			let source = T::ConvertAccountId::convert_account_id(sender.clone());
+			let gas_price = T::FeeCalculator::gas_price();
+
+			let vicinity = Vicinity {
+				gas_price,
+				origin: source,
+			};
+
+			let mut backend = Backend::<T>::new(&vicinity);
+			let mut executor = StackExecutor::new_with_precompile(
+				&backend,
+				gas_limit as usize,
+				&backend::GASOMETER_CONFIG,
+				T::Precompiles::execute,
+			);
+
+			let total_fee = gas_price * U256::from(gas_limit);
+			if Accounts::get(&source).balance < value + total_fee {
```

Comment:
is this needed if `executor.withdraw` will fail when balance is not enough anyway?

Reply:
Line below this withdraws the total fee, which fails if `balance` is lower than `total_fee`. However, the check for value, if without this line, only happens inside call, after all the substate structures are created. That is due to the behavior of `CALL` in Ethereum -- it adds stipend first into gas counter (which is considered even when the call failed), and only after then transfer condition is checked.

I'm fine removing this or keeping this, but currently I think keeping this would be slightly clearer. I do think there may be ways to refactor the stipend logic so that things can be nicer.